### DISCUSSION TITLE
RecursionError on deepcopy of Request

### DISCUSSION THREAD

**Discussion by mlegner** [encode/django-rest-framework#8682] [Potential Issue]
I noticed that calling `deepcopy` on a `Request` object causes infinite recursion with a `RecursionError`.

Minimal working example:

```python
import copy
from django.conf import settings
from django.http.request import HttpRequest
from rest_framework.request import Request

settings.configure()
request = Request(HttpRequest())
copy.deepcopy(request)
```

This produces the following output:

```text
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~/.asdf/installs/python/3.10.5/lib/python3.10/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "~/.asdf/installs/python/3.10.5/lib/python3.10/copy.py", line 272, in _reconstruct
    if hasattr(y, '__setstate__'):
  File "~/python_playground/.venv/lib/python3.10/site-packages/rest_framework/request.py", line 426, in __getattr__
    return getattr(self._request, attr)
  File "~/python_playground/.venv/lib/python3.10/site-packages/rest_framework/request.py", line 426, in __getattr__
    return getattr(self._request, attr)
  File "~/python_playground/.venv/lib/python3.10/site-packages/rest_framework/request.py", line 426, in __getattr__
    return getattr(self._request, attr)
  [Previous line repeated 993 more times]
RecursionError: maximum recursion depth exceeded
```

The problem seems to be the custom `__getattr__` function defined in [rest_framework/request.py#L410-L418](https://github.com/encode/django-rest-framework/blob/c10f2266222c434485889b08cc1463acdb8fa169/rest_framework/request.py#L410-L418):

```python
    def __getattr__(self, attr):
        """
        If an attribute does not exist on this instance, then we also attempt
        to proxy it to the underlying HttpRequest object.
        """
        try:
            return getattr(self._request, attr)
        except AttributeError:
            return self.__getattribute__(attr)
```

For some reason, there is a `Request` instance without `_request` attribute, so that the `self._request` in this function causes another call to `__getattr__`. A possible fix that works for me (and is otherwise equal to the current implementation) is the following:

```diff
-            return getattr(self._request, attr)
+            _request = self.__getattribute__("_request")
+            return getattr(_request, attr)
```

I would be happy to convert this to an issue and/or create a PR with this change.

My program versions:
- Python: 3.10.5
- Django: 4.1.2
- rest_framework: 3.14.0

**Comment by lovelydinosaur** [comment]
Okay - I don't *quite* understand that, but must be something to do with the implementation details of `deepcopy`.

(Ie. at some point we've got a new `Request` instance, but `._request` is not yet populated.)

I guess a test case and a fix would probably be okay here. What's the use-case for `deepcopy` on the request instance?

**Reply by mlegner** [reply]
Thanks for the quick feedback, @tomchristie! 

> Okay - I don't _quite_ understand that, but must be something to do with the implementation details of `deepcopy`.
> 
> (Ie. at some point we've got a new `Request` instance, but `._request` is not yet populated.)

Yes, that's what I found in my debugging session. But I also didn't quite get to the root cause there.

> What's the use-case for `deepcopy` on the request instance?

This is in the context of drf-spectacular, where `Request`s are attached to `Serializer`s for schema generation (see https://github.com/tfranzel/drf-spectacular/pull/792); when calling [`get_fields`](https://github.com/encode/django-rest-framework/blob/c10f2266222c434485889b08cc1463acdb8fa169/rest_framework/serializers.py#L381) and one of the fields is a serializer with a `Request` attached, this issue is triggered.

This does not happen with the default behavior of drf-spectacular, but I hit this with a custom `AutoSchema` in one of my projects.

 > I guess a test case and a fix would probably be okay here.

Should I create a PR for this?

**Reply by lovelydinosaur** [reply]
> Should I create a PR for this?

Sure thing, yup.

**Reply by lovelydinosaur** [reply]
---> https://github.com/encode/django-rest-framework/pull/8684
